x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/QueryableBuiltInRolesSynchronizer.java

```diff
@@ -0,0 +1,488 @@
+            || (cause instanceof ElasticsearchException
+                && "Failed to mark built-in roles as synced. The expected roles digests have changed.".equals(cause.getMessage()));
+    }
+
+    private boolean shouldSyncBuiltInRoles(ClusterState state) {
+        if (false == state.nodes().isLocalNodeElectedMaster()) {
+            logger.trace("Local node is not the master, skipping built-in roles synchronization");
+            return false;
+        }
+        if (false == state.clusterRecovered()) {
+            logger.trace("Cluster state has not recovered yet, skipping built-in roles synchronization");
+            return false;
+        }
+        if (queryableBuiltInRolesStore.isEnabled() == false) {
+            logger.trace("Role store is not enabled, skipping built-in roles synchronization");
+            return false;
+        }
+        if (state.nodes().getDataNodes().isEmpty()) {
+            logger.trace("No data nodes in the cluster, skipping built-in roles synchronization");
+            return false;
+        }
+        if (state.nodes().isMixedVersionCluster()) {
+            // To keep things simple and avoid potential overwrites with an older version of built-in roles,
+            // we only sync built-in roles if all nodes are on the same version.
+            logger.trace("Not all nodes are on the same version, skipping built-in roles synchronization");
+            return false;
+        }
+        if (false == featureService.clusterHasFeature(state, QUERYABLE_BUILT_IN_ROLES_FEATURE)) {
+            logger.trace("Not all nodes support queryable built-in roles feature, skipping built-in roles synchronization");
+            return false;
+        }
+        if (securityIndexDeleted) {
+            logger.trace("Security index has been deleted, skipping built-in roles synchronization");
+            return false;
+        }
+        if (isSecurityIndexClosed(state)) {
+            logger.trace("Security index is closed, skipping built-in roles synchronization");
+            return false;
+        }
+        return true;
+    }
+
+    private void doSyncBuiltinRoles(Map<String, String> indexedRolesDigests, QueryableBuiltInRoles roles, ActionListener<Void> listener) {
+        securityIndex.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+            final SecurityIndexManager frozenSecurityIndex = securityIndex.defensiveCopy();
+            if (frozenSecurityIndex.isAvailable(SecurityIndexManager.Availability.PRIMARY_SHARDS) == false) {
+                listener.onFailure(frozenSecurityIndex.getUnavailableReason(SecurityIndexManager.Availability.PRIMARY_SHARDS));
+
+            } else {
+                final Collection<RoleDescriptor> rolesToUpsert = rolesToUpsert(roles, indexedRolesDigests);
+                indexRoles(rolesToUpsert, frozenSecurityIndex, ActionListener.wrap(onResponse -> {
+                    final Set<String> rolesToDelete = rolesToDelete(roles, indexedRolesDigests);
```

Comment:
Nit: I'd pull this up to sit next to `rolesToUpsert` and assert that intersection of roles to delete and roles to update is empty, just as a sanity check